### ISSUE TITLE
Crash when using shortcuts to samba shares with $ in the name?

### ISSUE THREAD

**Issue by IvanRaide** [chrizonix/QuickLaunch#1]
Whenever I point QuickLauncher to a directory with shortcuts that have $ in the name it crashes on the regex interpretation of them.  Pasted the error below:

See the end of this message for details on invoking 
just-in-time (JIT) debugging instead of this dialog box.

************** Exception Text **************
System.FormatException: Invalid line 11 in ini file 'config.ini': 'DEV - c$.lnk = DEV - c$' key part 'DEV - c$.lnk' did not pass the regex validation.
   at GitHub.SimpleIni.IniFile.ParseLines(String[] lines, IniConfig config)
   at GitHub.SimpleIni.IniFile..ctor(String path, Nullable`1 config)
   at QuickLaunch.Data.ConfigFile.LoadFile()
   at QuickLaunch.Main.Main_Load(Object sender, EventArgs e)
   at System.Windows.Forms.Form.OnLoad(EventArgs e)
   at System.Windows.Forms.Form.OnCreateControl()
   at System.Windows.Forms.Control.CreateControl(Boolean fIgnoreVisible)
   at System.Windows.Forms.Control.CreateControl()
   at System.Windows.Forms.Control.WmShowWindow(Message& m)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.Form.WmShowWindow(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)

************** Loaded Assemblies **************
mscorlib
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9282.0 built by: NET481REL1LAST_C
    CodeBase: file:///C:/Windows/Microsoft.NET/Framework64/v4.0.30319/mscorlib.dll
----------------------------------------
QuickLaunch
    Assembly Version: 1.2.8.0
    Win32 Version: 1.2.8.0
    CodeBase: file:///C:/Applications/Enhancements/QuickLaunch/Shares/QuickLaunch.exe
----------------------------------------
System.Windows.Forms
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9256.0 built by: NET481REL1LAST_B
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Windows.Forms/v4.0_4.0.0.0__b77a5c561934e089/System.Windows.Forms.dll
----------------------------------------
System
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9282.0 built by: NET481REL1LAST_C
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System/v4.0_4.0.0.0__b77a5c561934e089/System.dll
----------------------------------------
System.Drawing
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9032.0 built by: NET481REL1
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Drawing/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Drawing.dll
----------------------------------------
AnySerializer
    Assembly Version: 1.2.0.0
    Win32 Version: 1.2.8.0
    CodeBase: file:///C:/Applications/Enhancements/QuickLaunch/Shares/QuickLaunch.exe
----------------------------------------
TypeSupport
    Assembly Version: 1.2.0.0
    Win32 Version: 1.2.8.0
    CodeBase: file:///C:/Applications/Enhancements/QuickLaunch/Shares/QuickLaunch.exe
----------------------------------------
System.Core
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9277.0 built by: NET481REL1LAST_B
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Core/v4.0_4.0.0.0__b77a5c561934e089/System.Core.dll
----------------------------------------
System.Configuration
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9032.0 built by: NET481REL1
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Configuration/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Configuration.dll
----------------------------------------
System.Xml
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9032.0 built by: NET481REL1
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Xml/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.dll
----------------------------------------

************** JIT Debugging **************
To enable just-in-time (JIT) debugging, the .config file for this
application or computer (machine.config) must have the
jitDebugging value set in the system.windows.forms section.
The application must also be compiled with debugging
enabled.

For example:

<configuration>
    <system.windows.forms jitDebugging="true" />
</configuration>

When JIT debugging is enabled, any unhandled exception
will be sent to the JIT debugger registered on the computer
rather than be handled by this dialog box.

**Comment by chrizonix** [comment]
Hi @IvanRaide, thanks for the report, I will have a look at it in a few days :)

**Comment by chrizonix** [comment]
Hi @IvanRaide, thanks for the report. Fixed with commit be3f8c74effb03aef6f70ceb59aae4784028cb13 in Release v1.2.8.2 :)
Wish you Merry Christmas and a Happy New Year!

**Comment by IvanRaide** [comment]
Thank you very much,

Merry Christmas and Happy New Year to you.

**Comment by IvanRaide** [comment]
Sorry, but now that it can walk those shares it looks like it hit some other regex validation problems.  I didnt create a new issue, but if needs I can.  Here is the info, (my guess is the @ sign is the problem)

See the end of this message for details on invoking 
just-in-time (JIT) debugging instead of this dialog box.

************** Exception Text **************
System.FormatException: Invalid line 401 in ini file 'config.ini': 'Sittu\ParioTech - original template files\Jekas\html Template Files\js\fancybox\fancybox_loading@2x.gif = fancybox_loading@2x' key part 'Sittu\ParioTech - original template files\Jekas\html Template Files\js\fancybox\fancybox_loading@2x.gif' did not pass the regex validation.
   at GitHub.SimpleIni.IniFile.ParseLines(String[] lines, IniConfig config)
   at GitHub.SimpleIni.IniFile..ctor(String path, Nullable`1 config)
   at QuickLaunch.Data.ConfigFile.LoadFile()
   at QuickLaunch.Main.Main_Load(Object sender, EventArgs e)
   at System.Windows.Forms.Form.OnLoad(EventArgs e)
   at System.Windows.Forms.Form.OnCreateControl()
   at System.Windows.Forms.Control.CreateControl(Boolean fIgnoreVisible)
   at System.Windows.Forms.Control.CreateControl()
   at System.Windows.Forms.Control.WmShowWindow(Message& m)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.Form.WmShowWindow(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)


************** Loaded Assemblies **************
mscorlib
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9282.0 built by: NET481REL1LAST_C
    CodeBase: file:///C:/Windows/Microsoft.NET/Framework64/v4.0.30319/mscorlib.dll
----------------------------------------
QuickLaunch
    Assembly Version: 1.2.8.2
    Win32 Version: 1.2.8.2
    CodeBase: file:///C:/Applications/Enhancements/QuickLaunch/Shares/QuickLaunch.exe
----------------------------------------
System.Windows.Forms
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9256.0 built by: NET481REL1LAST_B
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Windows.Forms/v4.0_4.0.0.0__b77a5c561934e089/System.Windows.Forms.dll
----------------------------------------
System
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9282.0 built by: NET481REL1LAST_C
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System/v4.0_4.0.0.0__b77a5c561934e089/System.dll
----------------------------------------
System.Drawing
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9032.0 built by: NET481REL1
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Drawing/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Drawing.dll
----------------------------------------
AnySerializer
    Assembly Version: 1.2.0.0
    Win32 Version: 1.2.8.2
    CodeBase: file:///C:/Applications/Enhancements/QuickLaunch/Shares/QuickLaunch.exe
----------------------------------------
TypeSupport
    Assembly Version: 1.2.0.0
    Win32 Version: 1.2.8.2
    CodeBase: file:///C:/Applications/Enhancements/QuickLaunch/Shares/QuickLaunch.exe
----------------------------------------
System.Core
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9277.0 built by: NET481REL1LAST_B
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Core/v4.0_4.0.0.0__b77a5c561934e089/System.Core.dll
----------------------------------------
System.Configuration
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9032.0 built by: NET481REL1
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Configuration/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Configuration.dll
----------------------------------------
System.Xml
    Assembly Version: 4.0.0.0
    Win32 Version: 4.8.9032.0 built by: NET481REL1
    CodeBase: file:///C:/Windows/Microsoft.Net/assembly/GAC_MSIL/System.Xml/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.dll
----------------------------------------

************** JIT Debugging **************
To enable just-in-time (JIT) debugging, the .config file for this
application or computer (machine.config) must have the
jitDebugging value set in the system.windows.forms section.
The application must also be compiled with debugging
enabled.

For example:

<configuration>
    <system.windows.forms jitDebugging="true" />
</configuration>

When JIT debugging is enabled, any unhandled exception
will be sent to the JIT debugger registered on the computer
rather than be handled by this dialog box.

**Comment by chrizonix** [comment]
Hi @IvanRaide, this is because the "@" symbol is not added to the list of valid characters. I can add this symbol tomorrow and you can test again :)

Just a side note. It looks like you are trying to add a full directory (with many many sub directories, and files) to the QuickLaunch menu. The problem with this approach is, that if you have thousand files in your folder, you will have thousand entries in the context menu.

So in your example, the following file `fancybox_loading@2x.gif` would be visible in the QuickLaunch toolbar as a single entry in the context menu:
`Sittu\ParioTech - original template files\Jekas\html Template Files\js\fancybox\fancybox_loading@2x.gif`

This makes the Context Menu very very slow, because Windows has to add/render thousand of Context Menus and Sub Context Menus. Do you really want this? Normally a QuickLaunch Toolbar is used to quickly launch a link to a folder or drive.

So instead of adding a complete folder (with all sub folders) you should consider creating a LINK to the folder, so in the Context Menu you would only have a link to the folder (with a single entry) in the context menu.

In this example I created three entries:
* LINK to Samba drive "c$ (127.0.0.1)"
* FULL folder QuickLaunch with all Sub Directories
* LINK to folder QuickLaunch (single entry)
![image](https://github.com/user-attachments/assets/f438b8f4-abce-4251-8759-d7adbf04f7f9)

You can also see it in the config:
* The LINK is a SINGLE entry, where as
* The FOLDER is stored as hundred entries
![image](https://github.com/user-attachments/assets/b83aaedd-8261-436a-a832-e1a01a4244da)

Please keep in mind, that I'm always building the FULL context menu before showing it to the user.
So the more entries you have, the slower the QuickLaunch will get.

I will release an update tomorrow regarding the "@" symbol, but I think adding all files and sub folders will be a bad user experience for you, since the QuickLaunch will be very slow :)

**Comment by chrizonix** [comment]
@IvanRaide I re-opened the ticket, so no worries :)
I will have a look tomorrow :)

**Comment by IvanRaide** [comment]
Actually that was a mistake, the directory Shares was supposed to have a link to Sittu, (not have the actual directory in it). I have corrected that, and its not crashing now (since its not walking the files of that Sittu folder and running into @ filenames.) so its working but I guess since could use the validation  Thanks for this quick fixes

**Comment by chrizonix** [comment]
Hi @IvanRaide, thanks for testing! I have now changed the logic for the "File Name Validator" from a whitelist, to a blacklist. So by default ALL characters are allowed (except for the "=" sign, because it is used in the config as a separator), but even if the "=" sign is used in filenames, the tool still fully works (the entry is recognized, and you can even launch the entry), only thing that doesn't work is, that during a refresh with the "Create/Load Config" button, the filename with the "=" sign is recognized as a new file, and hence will be moved to the back of the list (but manual sorting is still possible).

So long story short, now ALL filenames that are possible in Windows, will also work with this tool :)

Fixed in Commit: 536ac4bf7ecda6bb6035eea6dbe888432cc95cc4
Fixed in Release: v1.2.8.4

Again thank you very much for testing!
